redfish-core/lib/power_supply_metrics.hpp

```diff
@@ -0,0 +1,260 @@
+                                {
+                                    BMCWEB_LOG_DEBUG << "DBUS response error";
+                                    messages::internalError(aResp->res);
+                                    return;
+                                }
+
+                                const std::vector<
+                                    std::tuple<std::uint64_t, std::int64_t>>*
+                                    values_ptr =
+                                        std::get_if<std::vector<std::tuple<
+                                            std::uint64_t, std::int64_t>>>(
+                                            &values_rsp);
+                                if (values_ptr != nullptr)
+                                {
+                                    nlohmann::json& maximums =
+                                        aResp->res
+                                            .jsonValue["Oem"]["IBM"]
+                                                      ["InputPowerHistoryItem"]
+                                                      ["Max"];
+                                    for (const auto& values : *values_ptr)
+                                    {
+                                        // The first value returned is the
+                                        // timestamp, already handled in
+                                        // getAverageValues().
+                                        // The second value returned is in
+                                        // watts. It is the maximum watts this
+                                        // power supply has used in a 30 second
+                                        // interval.
+                                        maximums.push_back(std::get<1>(values));
+                                    }
+                                }
+                                else
+                                {
+                                    BMCWEB_LOG_ERROR
+                                        << "Failed to find power supply input "
+                                           "history Values data for:"
+                                        << powerSupplyID;
+                                }
+                            },
+                            serviceName, maximumPath,
+                            "org.freedesktop.DBus.Properties", "Get", interface,
+                            "Values");
+                    }
+                }
+            }
+        },
+        "xyz.openbmc_project.ObjectMapper",
+        "/xyz/openbmc_project/object_mapper",
+        "xyz.openbmc_project.ObjectMapper", "GetObject", maximumPath,
+        std::array<const char*, 1>{
+            "org.open_power.Sensor.Aggregation.History.Maximum"});
+}
+
+/**
+ * Systems derived class for delivering OemPowerSupplyMetric Schema.
+ */
+inline void requestRoutesPowerSupplyMetrics(App& app)
+{
+    BMCWEB_ROUTE(app, "/redfish/v1/Chassis/<str>/PowerSubsystem/"
```

Comment:
I will attempt to address issue #1.

Reply:
Isn't this already verified to be valid before we get here?
redfish-core/lib/power_supply.hpp:663 - getValidChassisID()
redfish-core/lib/power_supply.hpp:659 - getValidPowerSupplyID()

Reply:
My understanding is that each Redfish query provides a single level of information.  It does not recurse and provide information on levels below that automatically.

The file power_supply.hpp responds to the URI `/redfish/v1/Chassis/<str>/PowerSubsystem/PowerSupplies/<str>/`.

The file power_supply_metrics.hpp responds to the URI `/redfish/v1/Chassis/<str>/PowerSubsystem/PowerSupplies/<str>/Metrics`.

Since this is one level deeper, it is not running any of the code in power_supply.hpp, so you need to verify the chassis.

You could test to be sure.  Just specify `/redfish/v1/Chassis/my_name_is_brandon/PowerSubsystem/PowerSupplies/<str>/Metrics` and see if it complains.

Reply:
It does not complain, but it does not seem right that it replies.
```
$ curl -k -H "X-Auth-Token: ${BMCTOKEN}" -X GET https://${BMCHOST}:${BMCHTTPS}/redfish/v1/Chassis/chassis/PowerSubsystem/PowerSupplies/powersupply0/Metrics
{
  "@odata.id": "/redfish/v1/Chassis/chassis/PowerSubsystem/PowerSupplies/powersupply0/Metrics",
  "@odata.type": "#PowerSupplyMetric.v1_0_0.PowerSupplyMetric",
  "Id": "Metrics",
  "Name": "Metrics for powersupply0",
  "Oem": {
    "@odata.type": "#OemPowerSupplyMetric.Oem",
    "IBM": {
      "@odata.type": "#OemPowerSupplyMetric.IBM",
      "InputPowerHistoryItem": {
        "Average": [
          -6
        ],
        "Date": [
          "2022-05-05T15:25:39+00:00"
        ],
        "Max": [
          -5
        ]
      }
    }
  }
}
```
That is normal, expected.

```
$ curl -k -H "X-Auth-Token: ${BMCTOKEN}" -X GET https://${BMCHOST}:${BMCHTTPS}/redfish/v1/Chassis/chassis2/PowerSubsystem/PowerSupplies/powersupply0/Metrics
{
  "@odata.id": "/redfish/v1/Chassis/chassis2/PowerSubsystem/PowerSupplies/powersupply0/Metrics",
  "@odata.type": "#PowerSupplyMetric.v1_0_0.PowerSupplyMetric",
  "Id": "Metrics",
  "Name": "Metrics for powersupply0",
  "Oem": {
    "@odata.type": "#OemPowerSupplyMetric.Oem",
    "IBM": {
      "@odata.type": "#OemPowerSupplyMetric.IBM",
      "InputPowerHistoryItem": {
        "Average": [
          -6
        ],
        "Date": [
          "2022-05-05T15:25:39+00:00"
        ],
        "Max": [
          -5
        ]
      }
    }
  }
}
```
There is no chassis2! I got the same data. Weird.

```
$ curl -k -H "X-Auth-Token: ${BMCTOKEN}" -X GET https://${BMCHOST}:${BMCHTTPS}/redfish/v1/Chassis/sissahc/PowerSubsystem/PowerSupplies/powersupply0/Metrics
{
  "@odata.id": "/redfish/v1/Chassis/sissahc/PowerSubsystem/PowerSupplies/powersupply0/Metrics",
  "@odata.type": "#PowerSupplyMetric.v1_0_0.PowerSupplyMetric",
  "Id": "Metrics",
  "Name": "Metrics for powersupply0",
  "Oem": {
    "@odata.type": "#OemPowerSupplyMetric.Oem",
    "IBM": {
      "@odata.type": "#OemPowerSupplyMetric.IBM",
      "InputPowerHistoryItem": {
        "Average": [
          -6
        ],
        "Date": [
          "2022-05-05T15:25:39+00:00"
        ],
        "Max": [
          -5
        ]
      }
    }
  }
}
```
Again, that is not expected.